File: eldoria/models/stat_block.py
"""The full physical/magical profile of any creature, NPC, or player."""
from __future__ import annotations

from dataclasses import dataclass, field

from eldoria.models.dice import DiceFormula
from eldoria.models.status_effect import StatusEffect


@dataclass(frozen=True)
class MagicEffect:
    """A magic effect that shifts one trait/skill up (buff) or down (curse)."""

    name: str
    affected_trait: str
    magnitude: int
    beneficial: bool


@dataclass(frozen=True)
class StatBlock:
    tier: int
    strength: int
    agility: int
    willpower: int
    max_health: int
    armor_class: int
    speed: int
    attack_bonus: int
    damage: DiceFormula
    magic_damage: DiceFormula | None
    magic_effect: MagicEffect | None
    worth: int
    status_resistances: frozenset[StatusEffect] = field(default_factory=frozenset)

    @staticmethod
    def modifier_of(score: int) -> int:
        """(score - 10) // 2, D&D-style ability modifier. Python's // already floors like Kotlin's floorDiv."""
        return (score - 10) // 2

File: core/src/commonMain/kotlin/eldoria/core/model/StatBlock.kt
package eldoria.core.model

import kotlinx.serialization.Serializable

/** A magic effect that shifts one trait/skill up (buff) or down (curse), e.g. on a legendary item or a hostile spellcaster's touch. */
@Serializable
data class MagicEffect(
    val name: String,
    val affectedTrait: String,
    val magnitude: Int,
    val beneficial: Boolean,
)

/**
 * The full physical/magical profile of any creature, NPC, or player: every
 * field here is produced by StatGenerator from d6/d8/d20 rolls, scaled by
 * difficulty tier (1..5). This is what combat, encounters, and shops read
 * from -- nothing is flavor-text-only.
 */
@Serializable
data class StatBlock(
    val tier: Int,
    val strength: Int,
    val agility: Int,
    val willpower: Int,
    val maxHealth: Int,
    val armorClass: Int,
    val speed: Int,
    val attackBonus: Int,
    val damage: DiceFormula,
    val magicDamage: DiceFormula?,
    val magicEffect: MagicEffect?,
    val worth: Int,
    /** Statuses this being shrugs off entirely -- see model/StatusEffect.kt. Empty by default; Phase 2's ported named enemies (goblins, stone golems, etc.) get real per-type resistance lists. */
    val statusResistances: Set<StatusEffect> = emptySet(),
) {
    companion object {
        // (score - 10).floorDiv(2): java.lang.Math.floorDiv is JVM-only, this
        // is the multiplatform-safe Kotlin stdlib equivalent (since 1.5).
        fun modifierOf(score: Int): Int = (score - 10).floorDiv(2)
    }
}
